linux/driver/osp-iio.c

```diff
@@ -0,0 +1,1203 @@
+		.num_channels = 0,
+		.info = &osp_sensor_info,
+		.usebuffer = 0,
+		.useevent = 1,
+	},
+	[SENSOR_PRESSURE] = {
+		.name = "osp_pressure",
+		.trigname = "osp_pressure",
+		.channels = NULL,
+		.num_channels = 0,
+		.info = &osp_sensor_info,
+		.usebuffer = 1,
+		.useevent = 0,
+	},
+}, prv_sensor[NUM_PRIVATE_SENSOR_TYPE] = {
+	[PSENSOR_ACCELEROMETER_RAW] = {
+		.name = "osp_accel_raw",
+		.trigname = "osp_accel_raw",
+		.channels = accel_channels,
+		.num_channels = ARRAY_SIZE(accel_channels),
+		.info = &osp_sensor_info,
+		.usebuffer = 1,
+		.useevent = 0,
+	},
+	[PSENSOR_ACCELEROMETER_UNCALIBRATED] = {
+		.name = "osp_uncal_accel",
+		.trigname = "osp_uncal_accel",
+		.channels = accel_channels,
+		.num_channels = ARRAY_SIZE(accel_channels),
+		.info = &osp_sensor_info,
+		.usebuffer = 1,
+		.useevent = 0,
+	},
+	[PSENSOR_CONTEXT_DEVICE_MOTION] = {
+		.name = "osp_context",
+		.trigname = "osp_context",
+		.usebuffer = 0,
+	},
+};
+
+/* ------- sysfs ------- */
+
+static ssize_t osp_iio_renable(struct device *dev,
+			struct device_attribute *attr,
+			char *buf)
+{
+	struct iio_dev *indio_dev = dev_get_drvdata(dev);
+	struct osp_iio_sensor *osp_sensor = iio_priv(indio_dev);
+	ssize_t len = 0;
+
+	len += sprintf(buf, "%i\n", osp_sensor->state);
+
+	return len;
+}
+
+static ssize_t osp_iio_wenable(struct device *dev,
+			struct device_attribute *attr,
+			const char *buf, size_t len)
+{
+	struct iio_dev *indio_dev = dev_get_drvdata(dev);
+	struct osp_iio_sensor *osp_sensor = iio_priv(indio_dev);
+	int en;
+
+	if (buf[0] == '1')
+		en = 1;
+	else
+		en = 0;
+
+	OSP_Sensor_State(osp_sensor->sensor, osp_sensor->private, en);
+	osp_sensor->state = en;
+
+	return len;
+}
+
+/* --------------------- */
+
+/**
+ * osp_sensor_read_raw() - data read function.
+ * @indio_dev:	the struct iio_dev associated with this device instance
+ * @chan:	the channel whose data is to be read
+ * @val:	first element of returned value (typically INT)
+ * @val2:	second element of returned value (typically MICRO)
+ * @mask:	what we actually want to read. 0 is the channel, everything else
+ *		is as per the info_mask in iio_chan_spec.
+ */
+static int osp_sensor_read_raw(struct iio_dev *indio_dev,
+			      struct iio_chan_spec const *chan,
+			      int *val,
+			      int *val2,
+			      long mask)
+{
+	struct osp_iio_sensor *osp_sensor = iio_priv(indio_dev);
+	struct osp_iio_sensor *st = osp_sensor;
+	int ret = -EINVAL;
+
+	mutex_lock(&st->lock);
+	switch (mask) {
+	case 0: /* magic value - channel value read */
+		switch (chan->type) {
+		case IIO_ACCEL:
+		case IIO_ANGL_VEL:
+		case IIO_MAGN:
+			switch (chan->scan_index) {
+			case axis_x:
+				*val = osp_sensor->data.xyz.x;
+				break;
+			case axis_y:
+				*val = osp_sensor->data.xyz.y;
+				break;
+			case axis_z:
+				*val = osp_sensor->data.xyz.z;
+				break;
+			default:
+				break;
+			}
+			ret = IIO_VAL_INT;
+			break;
+		
+		default:
+			break;
+		}
+		break;
+	default:
+		break;
+	}
```

Comment:
we are reading data if channel types is IIO_ACCEL, IIO_MAG, IIO_ANG_VELOCITY only. But we have devices created with channel types as IIO_INCLI , IIO_QUATERNION  IIO_PRESSURE etc. do we need to add data for this channels as well?